platform/default/src/mbgl/map/map_snapshotter.cpp

```diff
@@ -4,218 +4,302 @@
+    ForwardingRendererObserver(RendererObserver& delegate_)
+        : mailbox(std::make_shared<Mailbox>(*Scheduler::GetCurrent())), delegate(delegate_, mailbox) {}
 
-    void setStyleURL(std::string styleURL);
-    std::string getStyleURL() const;
+    ~ForwardingRendererObserver() { mailbox->close(); }
 
-    void setStyleJSON(std::string styleJSON);
-    std::string getStyleJSON() const;
+    void onInvalidate() override { delegate.invoke(&RendererObserver::onInvalidate); }
 
-    void setSize(Size);
-    Size getSize() const;
+    void onResourceError(std::exception_ptr err) override { delegate.invoke(&RendererObserver::onResourceError, err); }
 
-    void setCameraOptions(CameraOptions);
-    CameraOptions getCameraOptions() const;
+    void onDidFinishRenderingFrame(RenderMode mode, bool repaintNeeded, bool placementChanged) override {
+        delegate.invoke(&RendererObserver::onDidFinishRenderingFrame, mode, repaintNeeded, placementChanged);
+    }
 
-    void setRegion(LatLngBounds);
-    LatLngBounds getRegion() const;
+    void onStyleImageMissing(const std::string& image, StyleImageMissingCallback cb) override {
+        delegate.invoke(&RendererObserver::onStyleImageMissing, image, std::move(cb));
+    }
 
-    void snapshot(ActorRef<MapSnapshotter::Callback>);
+private:
+    std::shared_ptr<Mailbox> mailbox;
+    ActorRef<RendererObserver> delegate;
+};
+
+class SnapshotterRenderer final : public RendererObserver {
+public:
+    SnapshotterRenderer(Size size, float pixelRatio, optional<std::string> localFontFamily)
+        : frontend(size,
+                   pixelRatio,
+                   gfx::HeadlessBackend::SwapBehaviour::NoFlush,
+                   gfx::ContextMode::Unique,
+                   std::move(localFontFamily)) {}
+
+    void reset() {
+        hasPendingStillImageRequest = false;
+        frontend.reset();
+    }
+
+    void onInvalidate() override { rendererObserver->onInvalidate(); }
+
+    void onResourceError(std::exception_ptr err) override {
+        hasPendingStillImageRequest = false;
+        rendererObserver->onResourceError(err);
+    }
+
+    void onDidFinishRenderingFrame(RenderMode mode, bool repaintNeeded, bool placementChanged) override {
+        if (mode == RenderMode::Full && hasPendingStillImageRequest) {
+            stillImage = frontend.readStillImage();
+        }
+        rendererObserver->onDidFinishRenderingFrame(mode, repaintNeeded, placementChanged);
+    }
+
+    void setObserver(std::shared_ptr<RendererObserver> observer) {
+        assert(observer);
+        rendererObserver = std::move(observer);
+        frontend.setObserver(*this);
+    }
+
+    void update(std::shared_ptr<UpdateParameters> params) {
+        assert(params);
+        hasPendingStillImageRequest = params->stillImageRequest;
+        frontend.update(std::move(params));
+    }
+
+    void setSize(Size size) { frontend.setSize(size); }
+
+    PremultipliedImage readStillImage() { return std::move(stillImage); }
```

Comment:
`takeImage` (can be taken only once)?

Reply:
yes, one snapshot should produce 1 image that is moved out. Renamed to takeImage